Swap for Variables, l1norm & l2norm for vectors, RunSingleTest works with parallel tests

nilsdeppe opened the pull request:
## Proposed changes

A few changes and additions I encountered while working on subcell.

- I'm back porting the Blaze l1Norm and l2Norm functions since they're useful and getting newer versions of Blaze to compile with AVX-512 was a nightmare.
- While back porting I wanted `RunSingleTest` to work with an ActionTesting test. I figured out how to get that to work and am adding it here.
- I wanted to swap two Variables where one had wrapped tags of the other, so that is being added.
- I needed to use `StripeIterator` (or at least explored it a bit) and so that gets some minor cleanup.
- I was confused by what our flux lifting was doing and so added a clarifying comment.
- I wanted to stream out a `gsl::span`, so I'm adding a stream operator for that.
- I found the same issue with wrapped tags not being printed correctly in `Variables` that @wthrowe  had found, but also cleaned up the code there a bit so I'm just submitting the cleanup.

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Add swap functionality to Variables

nilsdeppe committed:
Add stream operator for gsl::span

nilsdeppe committed:
Simplify Variables operator<< implementation

nilsdeppe committed:
StripeIterator cleanup.

- noexcept constructor and functions
- Move operator++ to header file so it can be inlined more.

nilsdeppe committed:
Clarify what the DG flux lifting is doing.

nilsdeppe committed:
Add l1norm and l2norm for Blaze vectors

nilsdeppe committed:
Fix link dependencies of Pypp

nilsdeppe committed:
Support parallel tests in RunSingleTest

fmahebert commented on src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp:
```diff
@@ -42,6 +42,14 @@ auto lift_flux(Variables<tmpl::list<FluxTags...>> flux,
+  //   w_0 = 2 / ((N + 1) * N)
+  //   w_N = 2 / ((N + 1) * N)
+  // The negative sign comes from bringing the boundary correction over to the
+  // RHS of the equal sign, while the magnitude of the normal vector above
```
I suggest clarifying "the equal sign". I know what you mean, but someone who hasn't derived DG a few times might want an extra bit of clarification here.

nilsdeppe replied:
I agree, though am struggling to figure out a better way to say that. Do you have a suggestion? I could do something like have the following in parens `du/dt = - boundary_correction`.

fmahebert replied:
Yeah, I think that would be an improvement. Or even `du/dt = (volume terms) - (boundary correction)` to not omit the other important piece :P

nilsdeppe replied:
Pff, who needs volume terms :P

fmahebert commented on tests/Unit/RunSingleTest/CMakeLists.txt:
```diff
@@ -3,22 +3,48 @@
+# different implementation in the cpp file in order to not depend on all
+# the test libraries.
+#
+# In order to run the tests in a single file:
```
I am confused about...
1. how many tests are compiled. The name `RunSingleTest` implies just one test, but here you say "the tests in a single file", implying multiple tests.
2. how many files are compiled. Here you say "the tests in a single file", implying all the tests should come from one source file, but in the next line you say "the cpp files", implying multiple files.

Can this be clarified (throughout this comment)?

fmahebert reviewed (commented):
A few non-binding suggestions on documentation :)

moxcodes commented on src/DataStructures/Variables.hpp:
```diff
@@ -829,6 +829,16 @@ bool operator!=(const Variables<TagsList>& lhs,
 }
 
+template <typename... TagsLhs, typename... TagsRhs,
+          Requires<not std::is_same<tmpl::list<TagsLhs...>,
```
why not `cpp17::is_same_v`?

nilsdeppe replied:
We've had weird compiler errors with some GCC versions when using that with variables so that's why I'm avoiding it. I don't know if it'll be an issue, but it was rather annoying to have the random failures.

fmahebert replied:
IIRC the gcc ICEs in the variables code were coming from `flat_all_v`? Were there also issues from `is_same_v` that I've since forgotten?

nilsdeppe replied:
I thought there used to be some related to this, but maybe I'm confusing the two :man_shrugging:

wthrowe replied:
I think they were both happening.  GCC seems to generally hate this file.

moxcodes commented on src/Utilities/Gsl.hpp:
```diff
@@ -847,6 +847,22 @@ constexpr ElementType& at(span<ElementType, Extent> s,
+    os << *it;
+    ++it;
+    for (; it != t.cend(); ++it) {
+      os << ",";
```
[nitpick] I'd prefer a space after the commas for readability, provided it doesn't go against a previously established procedure.

nilsdeppe replied:
The STL stream operators don't have spaces, so this is consistent with them.

moxcodes commented on src/Utilities/PointerVector.hpp:
```diff
@@ -85,6 +85,191 @@ template <typename T>
+}
+
+template <typename VT, bool TF>
+inline decltype(auto) l2Norm(const DenseVector<VT, TF>& dv) noexcept {
```
I'd prefer this to not be operable on complex types, because the complex version will need an abs in the first slot or a different power operation in order to make sense. I'd be happy with it either being SFINAE here or disallowed via the blaze map traits. The type trait route might be preferable if you're trying to minimally change the code imported from Blaze.

nilsdeppe replied:
Seems reasonable. I'd prefer to have that be done as a trait if possible since I'd like to have Blaze behavior be version independent. Do you have a suggestion on how to do this with the traits?

moxcodes replied:
well, shoot. I thought I did, but looking at this more carefully, there isn't actually a blaze struct for the norm function -- usually the dispatch here ends up going through a map or some such internal blaze thing, which gets the 'result type' from the type trait struct.
Alright, my suggestion is to try the following, and if it doesn't work, just leave it as is. Feel free to just make this an issue and assign it to me if you want to get this merged quickly or are unsatisfied with the below solution.
Put in `ComplexDataVector.hpp` a partial template specialization to the helper struct:
```
namespace blaze {
template <typename Abs, typename Power>
struct DVecNormHelper<ComplexDataVector, Abs, Power> {
};
}
```
I think this should make the bool lookup fail on line 262 and therefore not compile l2norm of a ComplexDataVector. Not as nice as I had hoped...

nilsdeppe replied:
I had to do something slightly different. This works for `l2Norm(vector)` but fails for `l2Norm(vector * vector)` (and probably other expressions). If desired we could add other specializations that can maybe deal with this more generally.

moxcodes replied:
I'm happy with this for now, there might be some template tricks we can pull to select anything for which the `ResultType` is `ComplexDataVector` in that specialization. Thanks for putting this in.

moxcodes commented on tests/Unit/DataStructures/Test_DataVector.cpp:
```diff
@@ -93,6 +97,23 @@ void test_data_vector_unary_math() noexcept {
   // `Test_DataVectorBinaryOperations.cpp` in an effort to better parallelize
   // the build.
+
+  // Test l1Norm and l2Norm:
```
I'm not sure I like placing the direct test code in line in these aggregated functions. Unfortunately, I don't see a good generic way of testing these 'reduction' like operations any differently than you've written here, and it doesn't really fit in `VectorImplTestHelpers` because the current norms only make sense for `DataVector`s. 
For now, could you just make a separate function for `data_vector_reduction_tests` or `data_vector_norm_tests` elsewhere in this file, and call it either from this function, or  from the `SPECTRE_TEST_CASE`, just to maintain the segmentation?

nilsdeppe replied:
I'll add a separate function and call it from the test case. I agree that keeping it segmented is cleaner :)

moxcodes commented on tests/Unit/Pypp/CMakeLists.txt:
```diff
@@ -10,5 +10,5 @@ add_test_library(
   "Pypp"
   "${LIBRARY_SOURCES}"
-  "ErrorHandling"
+  "DataStructures;ErrorHandling"
```
is this distinct in CMake syntax from having them on different lines? If not, why prefer same line?

nilsdeppe replied:
Yes because we have a function that takes a list this argument and in CMake lists are delimited by `;` (:man_shrugging:). Does that make sense?

moxcodes reviewed (changes requested):
A few nitpicks and small changes I'd like to see in the vector norm stuff.

nilsdeppe commented:
I have addressed what I could. Please let me know if the comments are any clearer. @moxcodes I'm not sure how to forbid `ComplexDataVector`s from "working" with the norms code.

fmahebert commented on tests/Unit/RunSingleTest/CMakeLists.txt:
```diff
@@ -4,15 +4,24 @@
+# for testing purposes. It uses the RunTests infrastructure with a slightly
 # different implementation in the cpp file in order to not depend on all
-# the test libraries.
+# the test libraries. Note that while typically the `RunSingleTest
```
missing backtick after RunSingleTest

fmahebert commented on tests/Unit/RunSingleTest/CMakeLists.txt:
```diff
@@ -4,15 +4,24 @@
+# executable will be used for quick write/build/run cycles of a single test,
+# more than one test and source file can be compiled. The main use of
+# `RunSingleTest` is to decrease turnaround time when changing very low-level
+# code such as DataVector or certain parts of Utitilies that cause most or
```
typo in Utilities

fmahebert reviewed (commented):
I think the fixups have nicely clarified what I was looking for 👍. Thanks! Just some typoes.

nilsdeppe commented:
Pushed fixups for typos and for clang-tidy with the Blaze code.

wthrowe commented on src/NumericalAlgorithms/DiscontinuousGalerkin/LiftFlux.hpp:
```diff
@@ -42,6 +42,14 @@ auto lift_flux(Variables<tmpl::list<FluxTags...>> flux,
     -> Variables<tmpl::list<db::remove_tag_prefix<FluxTags>...>> {
   auto lift_factor = std::move(get(magnitude_of_face_normal));
+  // The LGL weights are:
+  //   w_i = 2 / ((N + 1) * N * P_{N}(xi_i))
```
I think `P_{N}(xi_i)` is supposed to be squared.

nilsdeppe replied:
Thanks, you're right. Good catch! I've squashed this in since it's fairly minor.

moxcodes commented:
fixups look good to me. I'm happy with a squash (regardless of which way you decide for the l2norm of complex troubles - you can go ahead and squash that in too, It'll be easy to find because it will be a different file).

nilsdeppe commented:
Squashed and fixed. Thanks for the quick and helpful reviews everyone. Please let me know if I missed something or there's anything else you'd like to see changed! :)
